### DISCUSSION TITLE
Full Recalculate and RollupFullBatchRecalculator issues

### DISCUSSION THREAD

**Discussion by randybarton24** [jamessimone/apex-rollup#467] [Q&A] [answered]
Hi @jamessimone, first off thank you for your great work on this, it's much appreciated.

We've been attempting to implement apex-rollup to replace our DLRS implementation, and come across some things that seems a little strange.  

We've created 2 custom metadata rollups as a test, and performed the full recalculate.  They are COUNT and DISTINCT COUNT rollups.  I've noticed however that many records are not calculated correctly after the full recalculate completed.  If I go into the record and use the recalculate button, it calculates correctly, but not sure why the original recalculate didn't work.  

Also, I've been monitoring our Apex Jobs, and I notice the RollupFullBatchRecalculator job running very frequent.  In fact, yesterday we almost hit our DailyAsync limit with this job being ~85% of the total.  I'm wondering if you could help me understand what we should expect as the number of normal executions for this job and if you have any ideas that we could look into to resolve.

For reference, we were on version 1.5.67 when the issue occurred.  Upgrading to the latest version to see if that helps.

**Comment by jamessimone** [comment]
@randybarton24 if the job is running too frequently, it sounds like the default batch size may be timing out and restarting. You should only have the number of batches involved in the initial full recalc plus the number of inserts/updates/deleted as the full contribution from Apex Rollup to your async limits each day.

Typically with issues like this, I ask for one of the Batch Apex logs generated during a full recalc with the Is Rollup Logging Enabled flag set to true, and with it set to false. (I like to ask for one without logging enabled because sometimes the logs get truncated when it is enabled). Then I get a flame chart going to analyze where the timeout is occurring and work with you to help optimize the process. Sometimes, that's as simple as lowering the Batch Chunk Size on your rollup control.

I think the operation timing out is also responsible for you not seeing all of the appropriate field updates. Don't worry, we will get this figured out!

**Reply by randybarton24** [reply]
Thanks @jamessimone 

here are the log files
[apex-07L3q00004TD3jQEAT.log](https://github.com/jamessimone/apex-rollup/files/12114720/apex-07L3q00004TD3jQEAT.log)
[apex-07L3q00004TD3a6EAD.log](https://github.com/jamessimone/apex-rollup/files/12114721/apex-07L3q00004TD3a6EAD.log)

**Reply by jamessimone** [reply]
@randybarton24 apologies for the delay in response. I'm on a business trip this week and have gotten a bit behind in responses. I just took a look through the log you sent over - it looks like the one ending in 6EAD was just from the frontend polling the batch job status while the full recalc went off, but the one ending in QEAT had lots of good info in it.

At the moment, it's not clear exactly where an error would be occurring - I used a flame chart analyzer to look at the total transaction time spent for the batch of 500 items, which I've pasted below. Some highlights:

* the entire Apex Rollup calculation takes 2.3 seconds
* after that, the updating of the parent records takes another ~3.6 seconds to complete with the other processes you have going on:

<img width="901" alt="apex-rollup-randy-batch-analysis-2" src="https://github.com/jamessimone/apex-rollup/assets/16430727/0e31864e-03ca-48b8-99bd-f5326f7718f1">

This is a breakdown of the most time-expensive parts of the whole operation:

<img width="912" alt="apex-rollup-randy-batch-analysis" src="https://github.com/jamessimone/apex-rollup/assets/16430727/9ff4b1ef-af4f-4936-af9a-cb1ae0b99700">

Given the fact that you've barely used 10% of the amount of time allocated to a transaction spawned from a batch process (60 seconds total), it seems like you've got plenty of breathing room. This also makes me less confident that the issue is related to the Apex Rollup batch full recalc timing out, as I don't see any reason for there to be a significant difference in processing time between two batch chunks with nearly 50 seconds left of processing time total for each one.

Unfortunately, that means that we're back to the drawing board as far as figuring out where and how these timeouts are occurring for you. It'd be great to confirm that there isn't some sort of pre-existing, endlessly spinning, rollup recalculation going on in the background consuming Async Jobs without doing the right thing; once you've done a full recalc once, you really shouldn't need to use it again unless you're adding new fields to be rolled up and need to stamp the pre-existing values correctly. I will continue to work with you to understand how this issue is occurring, and ensure everything is working perfectly for you.

**Reply by randybarton24** [reply]
Thanks @jamessimone, I appreciate you looking into the logs.  I'll give it another try, and monitor it for repeating batch jobs.  I'll let you know if I see it happen again.  It's helpful to know what I should be expecting as normal. 👍

**Reply by randybarton24** [reply]
@jamessimone still struggling to get the COUNT and COUNT_DISTINCT rollups to calculate correctly.  If we use the LWC button to recalculate one record at a time, it works correctly, but when it is performed via the RollupFullBatchRecalculator.. it isn't accurate.  But I might be going about this the wrong way.  In our most recent testing we've been using a scheduled flow to perform a full recalculate of a subset of the records, since a full recalc takes several hours because of the large record count.  Including some screenshots of what we're using as the flow settings.  Each time we run the flow, I see the RollupFullBatchRecalculator running, and the COUNT_DISTINCT field seems to be updating correctly, but the COUNT field just keeps adding to the count each time it runs.  Please let me know if you have any suggestions.

![image](https://github.com/jamessimone/apex-rollup/assets/8461377/3219452e-4dae-4bbd-a9b5-44124cf535bd)

**Reply by jamessimone** [reply]
@randybarton24 try setting the Is Full Record Set optional input variable to the global constant for true and see if that helps? It sounds like there's some kind of zero'ing issue that I may need to investigate specifically for COUNT in a full recalc, but that flag may help in the meantime.

I don't think you're doing something wrong, and I'm confident we can work together to get this working properly.

**Reply by randybarton24** [reply]
@jamessimone today we tried migrating DLRS rollups to Apex Rollups and it appears RollupFullBatchRecalculator has gotten into an infinite loop again somehow.  The only way I can prevent it is to disable the org custom setting.  Once I re-enable it, the batches start firing again.  We didn't trigger any full recalculate jobs today since it's assumed that the DLRS fields are already up-to-date
![image](https://github.com/jamessimone/apex-rollup/assets/8461377/0dab7ca3-9786-4289-b403-e1dd6b3f7678)

**Reply by jamessimone** [reply]
@randybarton24 I have the fix for that issue lined up for the next version - I also need to make one other change to fix the issue you were reporting here with the counts not properly resetting. If you downgrade to version 1.5.75, you won't have to deal with the full batch recalc running over and over again.

**Reply by randybarton24** [reply]
@jamessimone ok thank you.  I'm currently on 1.5.74.  I'll wait until the new fix is available and try again.  Thanks again for all your help.

**Reply by jamessimone** [reply]
@randybarton24 the new version has been released! Hopefully that addresses these issues for you.

**Reply by randybarton24** [reply]
thanks @jamessimone, much appreciated.  Unfortunately, it still seems to be firing the batch after updating to 1.5.77.  I have Nebula Logger plugin installed as well, and I'm seeing these warning logs, which I'm assuming might be related?  If I understand correctly, if limits get exceeded it pushes to a batch job, but shouldn't the batch job just process the records being modified, not the entire FullRecalc of the database? 
![image](https://github.com/jamessimone/apex-rollup/assets/8461377/65985f1c-4038-447e-8801-1f2aabdb0d86)
![image](https://github.com/jamessimone/apex-rollup/assets/8461377/9e6a55d8-cc1b-4d83-845a-9f20514e7f08)
![image](https://github.com/jamessimone/apex-rollup/assets/8461377/7de5bb18-ca70-4fab-81d2-843626c1d151)
![image](https://github.com/jamessimone/apex-rollup/assets/8461377/cf87e188-7fef-4f1c-b288-e2e4dbe4ea1a)

**Reply by jamessimone** [reply]
@randybarton24 firstly, I would make sure there aren't any batch full recalcs running. Have you always started from a clean slate in the Apex Jobs page? Make sure all of the preexisting jobs are aborted!

 It's not clear to me from the screenshots how this whole process is starting. For example:

Were you starting to recalculate product subscriptions? that `hasExceededQueryNumberLimit` is being driven by the `Max Query Rows` value of 50 in your third screenshot. It sounds like there's a timeout occurring there because too many queries are being consumed, but based on what you've been reporting (two rollups with COUNT,COUNT_DISTINCT), Apex Rollup should really only be querying a few times in the entire batch full recalc:

- once to count the children prior to starting a batch (but this shouldn't count)
- once to pull back all of the parent items (twice if each rollup is going to different parents)
- the _only_ other queries that would run will be the result of any automation on the parent items being updated

It might be useful to take a look at the log messages leading up to the ones that you've shown to get a better understanding of what's going on.

Another key piece of this: Apex Rollup, when updating parents, will also attempt to run the rollups associated with those parents. This shouldn't be an issue for you because you've indicated you're only testing with two rollups, but since I've mentioned other automations starting up due to the parents being updated, I thought it best to make that clear.

Looking forward to hearing a bit more about what's going on in the logs (as all queries are logged that are run by Apex Rollup) so that we can get to the bottom of what's happening here.

**Reply by randybarton24** [reply]
@jamessimone yes, I'm disabling the custom setting so that new jobs stop firing, and aborting everything.  Also, I did migrate our DLRS rollups yesterday, so we have more than just the original two that we were testing with initially.  And in fact, product subscription as the child record has quite a few rollups configured.  So my process to replicate this issue is to disable the custom setting, abort everything, wait 10-15 mins just to make sure no more jobs get triggered, and then when I re-enable the custom setting the batch full recalcs immediately start queueing up, presumably from the trigger operations happening from standard day-to-day operations.  We haven't inititated a full recalc ourselves.

**Reply by randybarton24** [reply]
@jamessimone also, here are the full log files for the logs in the previous screenshots 
[Log-082323_00Di0000000dHcrEAE.log](https://github.com/jamessimone/apex-rollup/files/12305463/Log-082323_00Di0000000dHcrEAE.log)
[Log-082324_00Di0000000dHcrEAE.log](https://github.com/jamessimone/apex-rollup/files/12305469/Log-082324_00Di0000000dHcrEAE.log)

**Reply by randybarton24** [reply]
@jamessimone one more log captured with logging enabled in the Rollup Control, just in case the added detail helps... 

[Log-082787_00Di0000000dHcrEAE.log](https://github.com/jamessimone/apex-rollup/files/12306223/Log-082787_00Di0000000dHcrEAE.log)

**Reply by jamessimone** [reply]
It's definitely concerning that a batch full recalc is getting created for a routine day-to-day operation. That should not be happening. I will take a look at the logs, hopefully they yield something.

**Reply by jamessimone** [reply]
@randybarton24 are some of these logs manually concatenated by you? I'm looking specifically at the log ending in `HcrEAE.log` - for instance, we go from:

* Trigger.ProductSubscriptions - no-op here, which seems to align with what _should_ be happening when a rollup field is not updated in a child update
* Trigger.Assets - suddenly we hop to here with 100 updated items - perhaps these are updated from the product subscriptions in bulk? At the same time, it doesn't seem like any of the Accounts get updated here, which leads us to ...
* Trigger.ProductSubscriptions - whoah, we're back here. Recursion somewhere? Is this adding up so far?
* When we finally get down to the message including `Rollup v1.5.77: deferred rollups remaining` , it looks like the very first rollup shown in the list (the LAST operation on Most Recent Number Of Free Months, from the looks of it) only has 45 "calc items" (the children) - this is to be expected as it shows that many of them get processed. However, all the other rollups included show 100 calc items (the product subscriptions), which makes it sound like a timeout occurred during the first rollup

All of this is to say - there's something going on here which seems unusual, and it doesn't make sense that a CPU timeout is occurring with this few number of items being processed. It might help to generate a log at the `FINE` or `DEBUG` level without rollup logging on - this is the sort of thing I can pass through a log analyzer/flame chart to understand where some of these slowdowns are occurring. Again, happy to help in any which way I can as we work to figure out what's going on here.

**Reply by randybarton24** [reply]
@jamessimone I'm getting these log files from clicking the "Download Log File" from the Nebula Logger record... I can try to capture a standard Salesforce log, but they're just harder to catch the specific log that has WARN/ERROR logs
![image](https://github.com/jamessimone/apex-rollup/assets/8461377/ad77e891-ee79-4b73-b2d9-4762a95983ec)

**Reply by jamessimone** [reply]
Got it. OK, so it's good to know there's no manual work on your end! Let me know if my other questions make sense as far as the workflow goes and how these updates are cascading. It may help for you to set the `Max Rollup Retries` value on your Org Default Rollup Control CMDT record to -1 or something in the meantime - that will prevent these endless jobs

**Reply by randybarton24** [reply]
@jamessimone it doesn't make sense to me that there are so many calc records... there definitely would be some cascading triggering happening, but for the most part, there shouldn't be more than 15-20 child subscriptions per Account/Asset.  Also, I noticed as I read through the logs that there are rollups that I wouldn't expect to be triggering at all.  For example, the "Account_Limelight_URL" rollup shows 100 calc items.  That rollup should in theory only be triggered when an Asset.Limelight_URL__c field changes based on the WHERE clause configured in the Rollup metadata... that field would typically not be changed on more than ~50 records in an entire day. 
![image](https://github.com/jamessimone/apex-rollup/assets/8461377/90aab6eb-d372-4e04-9048-8f55022d1a48)

**Reply by jamessimone** [reply]
I think we can both agree that there's something weird happening here. It still seems like there's a rollup stuck in an endless loop - have you tried using the Anonymous Apex script to kill off all the running versions from the wiki?

**Reply by randybarton24** [reply]
@jamessimone yeah, just tried again to be sure, and there were 0 records returned.  I re-enable the custom setting, and the jobs start queueing up again.  I asked the user that shows as "Submitted By" on the job, and it looks like it coincides with them changing an Opportunity stage to closed lost, so I'm assuming that's the operation that triggered the rollup process to begin.  That action should trigger just the "Account_ActiveOpportunity" Rollup to fire

![image](https://github.com/jamessimone/apex-rollup/assets/8461377/66e9ac84-6f52-4a62-82d7-d326481eae80)
![image](https://github.com/jamessimone/apex-rollup/assets/8461377/658a3659-cd34-40a1-a173-0a6172d2feda)
![image](https://github.com/jamessimone/apex-rollup/assets/8461377/eea492ef-8e75-4c05-b0dc-b85bdfb3e911)

**Reply by jamessimone** [reply]
The _only_ time something like that should kick off a batch calculation would be if the account associated with that opportunity had more than 50,000 other opportunities - in this case, it's legitimate for a rollup to be kicked off (but not a full recalc rollup) because for a FIRST-based rollup, the current opportunity being closed might _be_ the prior FIRST value on the parent. That rollup should only be for the account associated with that opportunity, though.

If you yourself were on that opportunity record and updated it (potentially to another closed lost status, if possible), I'd be curious to see the initial log generated - basically, the very first thing in the debug logs _after_ you update that record should be the synchronous part of Apex Rollup doing some preparation. It's there, in that log, that we'll see exactly how and why a full recalculation is starting up instead of the standard rollup that should be getting generated.

**Reply by randybarton24** [reply]
thanks @jamessimone I replicated the Closed Lost process on the same opportunity.  It did indeed fire off the full recalc job again.  I think I got the right log (standard Salesforce log)🤞 
[apex-07L3q00004T67nGEAR.log](https://github.com/jamessimone/apex-rollup/files/12307775/apex-07L3q00004T67nGEAR.log)

**Reply by jamessimone** [reply]
Damn @randybarton24, it looks like the Rollup Logging flag wasn't enabled when that log was captured 🙃

**Reply by randybarton24** [reply]
@jamessimone no worries, good practice 😂 here's the log with flag enabled... behavior is consistent, it queues 3 full recalc jobs each time I change to Closed Lost.

[apex-07L3q00004T6834EAB.log](https://github.com/jamessimone/apex-rollup/files/12307899/apex-07L3q00004T6834EAB.log)

**Reply by jamessimone** [reply]
Haha thanks! That helps. Still not entirely clear what the issue is, but there might be enough to go off of here. I will keep you updated.

**Reply by jamessimone** [reply]
@randybarton24 calling it quits here for the night. I think maybe I'll still need another log or two on this front. It looks like whatever's happening that is problematic is either coming from the queueable spawned by the opp being updated, or whatever happens when the account gets updated after that. I was hoping to see in the logs some indication of how a completely routine update is getting transformed into a batch full recalc - as there are log messages specifically for that - and, failing that, track down anything unusual happening, but haven't been successful on that front yet.

**Reply by jamessimone** [reply]
@randybarton24 another thing that just occurred to me, which might also help (in addition to setting the Max Rollup Retries to 0 or -1 as I suggested earlier in the thread): if the Org Default rollup control CMDT record has its Should Run As picklist updated to Synchronous Rollup, that will condense the async portion of the logs into a single log (like the one you sent most recently). Seeing _everything_ that's happening in one place might go a long way towards figuring this out.

As well, I meant to mention this earlier - is DLRS still running? I saw quite a few mentions of the trigger for it in the logs. It's been a while but I know a few people who've experienced issues when first migrating were having problems because the old DLRS rollups were still running. Could be a red herring in this case, but always worth narrowing the possibilities for what's wrong. Thank you for your time - I appreciate you working with me as we get this resolved.

**Reply by randybarton24** [reply]
@jamessimone thank you so much for your help, I really appreciate it.  I grabbed a couple more logs here.  Both of them are with control set to synchronous, and one is with Max Rollup Retries at 10(apex-07L3q00004T68C3EAJ) and the other -1(apex-07L3q00004T68ZpEAJ).  Setting it to -1 did not trigger the batch jobs, 10 did, so hopefully that will help to compare the two.  I do have DLRS still running to keep the fields accurate for processes that depend on them while we figure this out.  I did deactivate them initially when I did the migration but re-enabled them once I realized that I would need to disable apex-rollup.
[apex-07L3q00004T68ZpEAJ.log](https://github.com/jamessimone/apex-rollup/files/12308615/apex-07L3q00004T68ZpEAJ.log)
[apex-07L3q00004T68C3EAJ.log](https://github.com/jamessimone/apex-rollup/files/12308616/apex-07L3q00004T68C3EAJ.log)

**Reply by jamessimone** [reply]
@randybarton24 OK, now we're talking! I see what the issue is, and it stems with these two rollups:

* Account_Child_Account_Active_Customer
* Account_Child_Account_Prospect

They're operating on `Account.ParentId` but I need to make sure that field isn't null! That's why the batch full recalcs are starting - they're looking at _every_ account without a Parent Account, which isn't at all right. Now that I know what the issue is, I should be able to fix this easily.

**Reply by jamessimone** [reply]
@randybarton24 new version is rolling out as we speak. I will have limited availability over the next few days, but I will be monitoring this thread!

**Reply by randybarton24** [reply]
@jamessimone thank you so much! looking good so far!

**Reply by jamessimone** [reply]
Awesome! Love to hear it

**Reply by randybarton24** [reply]
@jamessimone we seem to be all good with the full batch recalcs not getting triggered with the latest version, thanks again for your help in getting that figured out.  I'm having some new issues now.  If you'd prefer I open them as a new discussions, let me know.

- Getting several "SObject row was retrieved via SOQL without querying the requested field" errors.  
- several errors of "No such column 'field' on entity 'Account'
- appears to be failing in our batch processing with a null pointer exception: `System.NullPointerException: Attempt to de-reference a null object Class.RollupAsyncProcessor.getAsyncRollup: line 379`

**Reply by jamessimone** [reply]
@randybarton24 the last bullet point is an easy fix thanks to that stacktrace. Any chance you have a log for the first two errors? I'll try to see if I can't get those errors to appear, myself.

**Reply by randybarton24** [reply]
@jamessimone unfortunately I haven't been able to get a log for those other two yet.  I had to disable the custom setting again because we had so many failures in our batch processing.  Am I correct in assuming that the hierarchy custom setting allows me to enable the package for specific users?  That way I could enable for a few users to capture a log.

**Reply by jamessimone** [reply]
@randybarton24 yes, that's correct. I will probably push up a fix for that null pointer exception today, and hopefully we can coordinate seeing where these other issues occur!

**Reply by randybarton24** [reply]
@jamessimone perfect thank you.  I think I'll be able to re-enable for the whole org once the null pointer is fixed.  I looked a little closer at the other errors, and I think it might be only happening on the rollups where Sharing Mode is set to User.  Don't have any logs yet though.

**Reply by jamessimone** [reply]
@randybarton24 no problem - the latest version should fix that null pointer issue.

**Reply by randybarton24** [reply]
thanks @jamessimone, I've installed the latest version, but still getting the null pointer exception.  Here's a full log. 
[apex-07L3q00004T6BcpEAF.log](https://github.com/jamessimone/apex-rollup/files/12371715/apex-07L3q00004T6BcpEAF.log)

**Reply by jamessimone** [reply]
@randybarton24 ah, I see where that's coming from now. Apologies for that oversight - rebuilding a new package now.

**Reply by jamessimone** [reply]
OK, apologies again for that false start - with 1.5.81 you should _actually_ be good to go

**Reply by randybarton24** [reply]
@jamessimone awesome, thank you!  I'll let you know if I'm able to get any logs on the other errors.  I switched all rollups to be System Sharing Mode, so we'll see if that does it.

**Reply by randybarton24** [reply]
@jamessimone I was able to catch a log on the "SObject row was retrieved via SOQL without querying the requested field" error.
[apex-07L3q00004T6CzlEAF.log](https://github.com/jamessimone/apex-rollup/files/12380980/apex-07L3q00004T6CzlEAF.log)

**Reply by jamessimone** [reply]
@randybarton24 OK I'm taking a look at that. Is this a full recalc that's been submitted through the full recalculation app? Or through some other means? I see there's about 52k product subscriptions rolling up to Asset. In particular, I think what I'm looking to discover is if there's any chance this full recalculation is starting from "the top" (so to speak) or if it's cascading from another operation. I have some theories as to what's happening, but making sure I understand what's going on will help to confirm them.

**Reply by randybarton24** [reply]
@jamessimone we haven't submitted any full recalc jobs, so everything is coming from cascading operations.

**Reply by jamessimone** [reply, marked as answer]
@randybarton24 I'm rolling out a new version which should address this issue. It was a bit difficult to reproduce the exact error you were getting, but I think based on the info you've supplied and the investigation I did that this is likely how the error was occurring.

**Reply by randybarton24** [reply]
thanks @jamessimone I appreciate you looking into it.  I'll install the newest version and let you know how it goes
